### ISSUE TITLE
Support "tests passed on rerun" on Azure DevOps

### ISSUE THREAD

**Issue by dgkanatsios** [ricardofslp/dotnet-trx-merge#67]
Hey there, thanks for the awesome tool, works like a charm!
We are using it on AzureDevOps to merge .trx files by doing `dotnet test` with retries (using a script very similar to this https://www.cazzulino.com/dotnet-test-retry.html).
We are merging the results into a single file which then gets uploaded via the [PublishTestResults](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/publish-test-results-v2) ADO task.

The only issue we see is that, normally, tests that failed but subsequently passed are reported as just "Passed" whereas they should be reported as "Passed on rerun". Any chance we can add this feature to this great tool? Much appreciated!

**Comment by ricardofslp** [comment]
Hi @dgkanatsios,

I know but I also have a use case where I need to actually have a pass instead of a passed in a rerun.
But thanks for this info.

I think it should be enough for me to work on this.

**Comment by ricardofslp** [comment]
Hi @dgkanatsios,

I'm trying to find a way to do this but this seems to be a trick done in the api.
At least I tried to change the state but I don't think passed on rerun is actually a state or do you have a trx file with a similar state?

**Comment by dgkanatsios** [comment]
hey @ricardofslp thanks again for checking for this.
Indeed, it doesn't seem to look like a state. I managed to find test results and see the .trx. They're gonna need some cleanup before I can share them but, in a test run where two tests passed on re-run, I can see 2 .trx files. The first .trx file contains data for all the tests and has the failed tests shown as "Failed"

```xml
  <UnitTestResult executionId="f2fa4d1e-97cb-44dd-9fba-d3fb3a99c664" testId="a787a6b2-be35-63ee-c497-74ec580998d2" testName="TESTNAME" computerName="..." duration="00:20:10.5257967" startTime="2023-12-02T01:16:30.7530123+00:00" endTime="2023-12-02T01:38:42.7892992+00:00" testType="13cdc9d9-ddb5-4fa4-a97d-d965ccfc6d4b" outcome="Failed" testListId="8c84fa94-04c1-424b-9868-57a2d4851a1d" relativeResultsDirectory="f2fa4d1e-97cb-44dd-9fba-d3fb3a99c664">
      <Output>
        <StdOut>...</StdOut>
        <ErrorInfo>
          <Message>...</Message>
          <StackTrace>   ...
</StackTrace>
        </ErrorInfo>
      </Output>
    </UnitTestResult>
...

  <ResultSummary outcome="Failed">
    <Counters total="245" executed="243" passed="241" failed="2" error="0" timeout="0" aborted="0" inconclusive="0" passedButRunAborted="0" notRunnable="0" notExecuted="0" disconnected="0" warning="0" completed="0" inProgress="0" pending="0" />
    <Output>
      <StdOut>...
</StdOut>
    </Output>
```

on the second .trx file, I can see that only the two tests that failed have executed

```xml
  <UnitTestResult executionId="db6b289d-9029-4c2e-899e-add811ecc822" testId="a787a6b2-be35-63ee-c497-74ec580998d2" testName="TESTNAME" computerName="..." duration="00:01:13.5733330" startTime="2023-12-02T01:54:11.4971182+00:00" endTime="2023-12-02T01:57:32.5289032+00:00" testType="13cdc9d9-ddb5-4fa4-a97d-d965ccfc6d4b" outcome="Passed" testListId="8c84fa94-04c1-424b-9868-57a2d4851a1d" relativeResultsDirectory="db6b289d-9029-4c2e-899e-add811ecc822">
      <Output>
        <StdOut>
...
 </Output>
    </UnitTestResult>
...
 <ResultSummary outcome="Completed">
    <Counters total="2" executed="2" passed="2" failed="0" error="0" timeout="0" aborted="0" inconclusive="0" passedButRunAborted="0" notRunnable="0" notExecuted="0" disconnected="0" warning="0" completed="0" inProgress="0" pending="0" />
    <Output>
      <StdOut>...
</StdOut>
    </Output>
```

TBH, not sure if this is enough or VsTest task does something else behind the scenes :(

Seems that this is a known issue https://stackoverflow.com/questions/70840866/how-to-update-azure-devops-test-results-with-to-display-passedonrerun

**Comment by ricardofslp** [comment]
I found that article too :( 
It seems the VsTest task does some magic behind the scenes probably with the api. This would make it hard to support this. 

I was also trying to see if we could add something to the tags column but not even that.

**Comment by ricardofslp** [comment]
Hi @dgkanatsios,

Being this some magic done on the azure devops task I would not like to have api calls happening inside this tool. Even so I would like to explore another option but this one may take a while to investigate.

Will come back with feedback on this issue whenever I have an answer.

**Comment by thenutter** [comment]
I'd also love to see this functionality.

--

For my use case, I'd also like the ability to drop (or replace) the failed test if it passes on a re-run.

**Comment by ricardofslp** [comment]
Hi @thenutter,

Regarding the replace it should be happening now. It just doesn't mark as a Pass on Rerun because that is something done by the task by calling the API.

**Comment by thenutter** [comment]
Thanks @ricardofslp, with the latest version when I pass the combined TRX to any reporting tool I see two separate runs, one pass and one fail (if passed on re-run). I'd expect if it passes on a re-run, I'd not see this within the report.

What reporting tools are people using?

**Comment by ricardofslp** [comment]
Hi @thenutter,

Could you check the merged trx and see if that test appears twice or only once in the results?

**Comment by thenutter** [comment]
@ricardofslp results appearing twice here, one failed then a pass. Although, the files not opening in visual studio is blocking me at the moment.

**Comment by ricardofslp** [comment]
Hi @thenutter,

Is it possible to  share the source trx files before the merge? 
Does this only happen in the latest version or happen also in the prior ones?

**Comment by ricardofslp** [comment]
Hi @thenutter ,

Whenever you reply could you open a new issue please?
This issue seems to be different than the original request

**Comment by martin-nutter** [comment]
Tested this again and it works 👍 only the latest result is showing.

Nice one @ricardofslp
